README
=== README for CPAN-Site version 1.06
=   Generated on Tue May 18 16:07:49 2010 by OODoc 1.06

There are various ways to install this module:

 (1) if you have a command-line, you can do:
       perl -MCPAN -e 'install <any package from this distribution>'

 (2) if you use Windows, have a look at http://ppm.activestate.com/

 (3) if you have downloaded this module manually (as root/administrator)
       gzip -d CPAN-Site-1.06.tar.gz
       tar -xf CPAN-Site-1.06.tar
       cd CPAN-Site-1.06
       perl Makefile.PL
       make          # optional
       make test     # optional
       make install

For usage, see the included manual-pages or
    http://search.cpan.org/dist/CPAN-Site-1.06/

Please report problems to
    http://rt.cpan.org/Dist/Display.html?Queue=CPAN-Site



ChangeLog

=== version history for CPAN::Site

   *** See the CPAN6 development at http://cpan6.org

version 1.06: Tue May 18 16:50:11 CEST 2010
	Fixes:
	- adapt to newer version.pm, which broke interface.
	  rt.cpan.org #57575 [Richard Huxton]

version 1.05: Tue Mar  2 09:33:18 CET 2010
	Improvements:
	- add --env_proxy
	  rt.cpan.org#55008 [Robert Bohne]

version 1.04: Mon Oct 26 14:56:57 CET 2009
	Improvements:
	- produce error, not info, if a package cannot be collected.
	- some informational messages from mirror and index processes
	  decreased to "debug" level (-vvv or --mode="DEBUG").
	- better attempt for VERSION parsing.
	- use age of index to determine whether an archive has to
	  be inspected again.

version 1.03: Wed Oct  7 19:46:57 CEST 2009
	Fixes:
	- use $file->full_path i.s.o $file->name for files in a tar,
	  required to correctly access some tar archives.
	  rt.cpan.org#49795 [Matisse Enzer] and [Alex Efros]
	- accept archives which end on .tgz
	- keep highest version number of a pm, not the last [Alex Efros]

version 1.02: Wed Sep  9 22:44:46 CEST 2009
	Fixes:
	- "version" action did not produce version
	- "mirror" did not initialize LWP::UserAgent
	- some modules, like DBI, specify package versions outside
	  the package scope.

	Improvements:
	- prefer new CPANSITE_LOCAL as alternative for CPANSITE
	  environment variable.

version 1.01: Wed Sep  2 12:55:19 CEST 2009
	Changes:
	- removed the need for client-side CPAN::Site installation.
	- update 02packages after 1 hour (was 1 day), because CPAN
	  is refreshing more often, lately.
	- removed [MYCPAN] argument to "index", because there is
	  a --site option now.
	- CPANSITE_CPAN renamed to CPANSITE_GLOBAL
	- option --stand-alone replaces --cpan ''

	Fixes:
	- require Log::Report 0.25, not the non-existent 0.27

	Improvements:
	- indexer synchronizes with core CPAN after 1 hour (was 1 day)
	  because CPAN is faster than previously.
	- CPANSITE_CPAN environment variable as alternative to --site
	- Replace IO::Compress::* by IO::Zlib, which is already required
	  by Archive::Tar.
	- explain changes in explain_101.txt
	- --cpan option alternative added: -c

version 1.00: Tue Sep  1 10:48:02 CEST 2009
	Large version number jump to indicate structural rewrite
	of internals.  This may trigger someones awareness to be
	more careful in the upgrade.

	Changes:
	- removed mkpackages which was kept for backward compatibility
	- do not extract readme's anymore, probably no-one uses them

	Improvements:
	- use Archive::Tar to unpack archives, to replace the tricky (but
	  fast) scanning of the tar binary.
	- replace use of external 'gzip' binary to IO::Compress
	- use File::Spec in stead of UNIX syntax. The module is probably
	  still not usable on Windows, but coming close.
	- use Log::Report for error/trace/verbose
	- rename t/CPAN-Site-Index.t into t/10index.t
	- improved trace on index
	- new 'mirror' option of cpansite

version 0.26: Mon Aug 31 09:51:55 CEST 2009

	Fixes:
	- do not use /bin/pwd in t/pod.t
	- accept http to retrieve data from central CPAN as alternative
          to ftp, by using LWP in stead of Net::FTP

	Improvements
	- clean-up/extend manual of CPAN::Site

version 0.25: Wed Jan 21 09:59:41 CET 2009

	Fixes:
	- also client requires HTTP::Date

	Improvements:
	- regressions tests for fix implemented in previous release.
	  [Matisse Enzer]

version 0.24: Tue Dec 30 11:36:39 CET 2008

	Fixes:
	- do only include package names found on usual locations.
	  rt.cpan.org#41935 [Matisse Enzer]

	Improvements:
	- speed-up

version 0.23: Tue Dec 23 13:30:25 CET 2008

	Fixes:
	- tests require Test::More 0.82 [cpantesters]

version 0.22: Mon Dec 22 15:26:11 CET 2008

	Fixes:
	- pre-req version.pm [Sander Hulst]

	Improvements:
	- regression test by [Matisse Enzer]
	- packages without version are enlisted [Matisse Enzer]
	- without version, then enlisted as undef

version 0.21: Thu May  8 13:15:22 CEST 2008

	Fixes:
	- use CPAN::HandleConfig if available.
	  rt.cpan.org#34869 [Alexander Kühne]

version 0.20: Thu Apr 10 12:13:26 CEST 2008
	- accept version objects [Alex Efros]
	- no warning "undef" when archive location is not specified
	- 01mailrc and 03modlist are updated from the real CPAN, not
	  just empty.  Motivated by contributions by [Alex Efros]
	- create directory trees when missing.  Only the top-level
	  must exist, to protect against errors.
	- Makefile.PL query about role not used [Alex Efros]
	- "Cpansite for the impatient" documentation by [Ales Efros]

version 0.19: Fri Mar 21 08:13:19 CET 2008
	- fix in distname [Matisse Enzer]

version 0.18: Wed Mar 19 18:35:37 CET 2008
	- cpansite script should strip file: with two slashes or none,
	  to permit absolute path.
	- fix when local archive is empty
	- only take pm's from lib- and top-dir, not f.i. from the
	  test-scripts [Matisse Enzer]

version 0.17: Wed Jan 30 16:44:07 CET 2008
	- fix for open file error [Matisse Enzer]
	- different way of defining the README output filename [Matisse Enzer]
	- minor output improvements. [Matisse Enzer]

version 0.16: Fri Nov  9 23:38:10 CET 2007
	- use package directory explicitly in the creation of the
	  readme extract. [Matisse Enzer]

version 0.15: Fri Jun  8 17:01:30 CEST 2007
	- document how to change urllist from CPAN.pm shell [Abe Timmerman]
	- removed short filename-support (MSDOS) from CPAN::Site
	- use empty 03modlist.data.gz
	- add t/pod.t

version 0.14: Tue Dec 19 13:42:48 CET 2006
	- use oodist (from OODoc) to produce manual pages
	- official new maintainer MARKOV
	- first public release of rewrite.

version 0.13:
	- depend on HTTP::Date if server
	- index generation code moved from bin/cpansite into
	  lib/CPAN/Site/Index.pm
	- Add CPANSITE_ROLE environment variable to Makefile.PL
	- show version of cpansite when shell is started

version 0.12:
	- 01mailrc is removed when the core cpan is merged in, and
	  created otherwise.

version 0.11:
	- use HTTP::Date to produce timestamp on package

version 0.10:

lib/CPAN/Site/Index.pm
# Copyrights 1998,2005-2010 by Mark Overmeer.
#  For other contributors see ChangeLog.
# See the manual pages for details on the licensing terms.
# Pod stripped from pm file by OODoc 1.06.
use warnings;
use strict;

package CPAN::Site::Index;
use vars '$VERSION';
$VERSION = '1.06';

use base 'Exporter';

our @EXPORT_OK = qw/cpan_index cpan_mirror/;
our $VERSION;  # required in test-env

use Log::Report     'cpan-site', syntax => 'SHORT';

use version;
use File::Find      qw/find/;
use File::Copy      qw/copy/;
use File::Basename  qw/basename dirname/;
use HTTP::Date      qw/time2str/;
use File::Spec::Functions qw/catfile catdir splitdir/;
use LWP::UserAgent  ();
use Archive::Tar    ();
use CPAN::Checksums ();
use IO::Zlib        ();

my $tar_gz      = qr/ \.tar\.gz$ | \.tar\.Z$ | \.tgz$/x;
my $cpan_update = 0.04; # days between reload of full CPAN index
my $ua;

sub safe_copy($$);
sub cpan_index($@);
sub register($$$);
sub package_inventory($$;$);
sub package_on_usual_location($);
sub inspect_archive;
sub collect_package_details($$);
sub update_global_cpan($$);
sub load_file($$);
sub merge_global_cpan($$$);
sub create_details($$$$$);
sub calculate_checksums($);
sub read_details($);
sub remove_expired_details($$$);
sub mkdirhier(@);
sub cpan_mirror($$$@);

sub safe_copy($$)
{   my ($from, $to) = @_;
    trace "copy $from to $to";
    copy $from, $to
        or fault __x"cannot copy {from} to {to}", from => $from, to => $to;
}

sub cpan_index($@)
{   my ($mycpan, $globalcpan, %opts) = @_;
    my $lazy     = $opts{lazy};
    my $fallback = $opts{fallback};
    my $undefs   = exists $opts{undefs} ? $opts{undefs} : 1;

    unless($ua)
    {   $ua = LWP::UserAgent->new;
        $ua->env_proxy if $opts{env_proxy};
    }

    -d $mycpan
        or error __x"archive top '{dir}' is not a directory"
             , dir => $mycpan;

    my $program     = basename $0;
    $VERSION      ||= 'undef';   # test env at home
    trace "$program version $VERSION";

    my $global      = catdir $mycpan, 'global';
    my $mods        = catdir $mycpan, 'modules';
    mkdirhier $global, $mods;

    my $globdetails = update_global_cpan $mycpan, $globalcpan;

    # Create mailrc and modlist

    safe_copy catfile($global, '01mailrc.txt.gz')
            , catfile($mycpan, authors => '01mailrc.txt.gz');

    safe_copy catfile($global, '03modlist.data.gz')
            , catfile($mods, '03modlist.data.gz');

    # Create packages details

    my $details     = catfile $mods, '02packages.details.txt.gz';
    my $newlist     = catfile $mods, '02packages.details.tmp.gz';
    my $newer;

    my $reuse_dists = {};
    if($lazy && -f $details)
    {   $reuse_dists = read_details $details;
        $newer       = -M $details;
        remove_expired_details $mycpan, $reuse_dists, $newer;
    }

    my ($mypkgs, $distdirs)
      = package_inventory $mycpan, $reuse_dists, $newer;

    merge_global_cpan $mycpan, $mypkgs, $globdetails
        if $fallback;

    create_details $details, $newlist, $mypkgs, $lazy, $undefs;

    if(-f $details)
    {   trace "backup old details file to $details.bak";
        safe_copy $details, "$details.bak";
    }

    if(-f $newlist)
    {   trace "promoting $newlist to current";
        rename $newlist, $details
            or error __x"cannot rename '{from}' in '{to}'"
                 , from => $newlist, to => $details;
    }

    calculate_checksums $distdirs;
}

#
# Package Inventory
#

# global variables for testing purposes (sorry)
our ($topdir, $findpkgs, %finddirs, $olddists, $index_age);

sub register($$$)
{  my ($package, $this_version, $dist) = @_;

# disabled: too verbose, even for for trace
#  trace "register $package, "
#      . (defined $this_version ? $this_version : 'undef');

   my $registered_version = $findpkgs->{$package}[0];

   $this_version =~ s/^v// if defined $this_version;
   return if !defined $this_version;
   return if defined $registered_version
          && qv($registered_version) > qv($this_version);
   # above with qv() works well, even with "undef" in the variables. See t/20qv

   $findpkgs->{$package} = [ $this_version, $dist ];
}

sub package_inventory($$;$)
{  (my $mycpan, $olddists, $index_age) = @_;   #!!! see "my"
   $topdir   = catdir $mycpan, 'authors', 'id';
   mkdirhier $topdir;

   $findpkgs = {};
   trace "creating inventory from $topdir";

   find {wanted => \&inspect_archive, no_chdir => 1}, $topdir;
   ($findpkgs, \%finddirs);
}

sub package_on_usual_location($)
{  my $file  = shift;
   my ($top, $subdir, @rest) = splitdir $file;
   defined $subdir or return 0;

      !@rest             # path is at top-level of distro
   || $subdir eq 'lib';  # inside lib
}

sub inspect_archive
{   my $fn   = $File::Find::name;
    -f $fn && $fn =~ $tar_gz
        or return;

    (my $dist = $fn) =~ s!^$topdir[\\/]!!;
    if(defined $index_age && -M $fn > $index_age)
    {
        unless(exists $olddists->{$dist})
        {   trace "not the latest: $dist";
            return;
        }

        trace "latest older than index: $dist";
        foreach (@{$olddists->{$dist}})
        {   my ($pkg, $version) = @$_;
            register $pkg, $version, $dist;
        }
        return;
    }

    trace "inspecting archive $fn";
    $finddirs{$File::Find::dir}++;

    my $arch =  Archive::Tar->new;
    $arch->read($fn, 1)
        or error __x"no files in archive '{fn}': {err}"
             , fn => $fn, err => $arch->error;

    foreach my $file ($arch->get_files)
    {   my $fn = $file->full_path;
        $file->is_file && $fn =~ m/\.pm$/i && package_on_usual_location $fn
            or next;
        collect_package_details $dist, $file;
    }
}

sub collect_package_details($$)
{   my ($dist, $file) = @_;

    my @lines  = split /\r?\n/, ${$file->get_content_by_ref};
    my $in_pod = 0;
    my $package;
    local $VERSION = undef;  # may get destroyed by eval

    foreach (@lines)
    {   last if m/^__(?:END|DATA)__$/;

        $in_pod = ($1 ne 'cut') if m/^=(\w+)/;
        next if $in_pod;

        if( m/^\s* package \s* ((?:\w+\:\:)*\w+) \s* ;/x )
        {   # second package in file?
            if(defined $package)
            {   my $nextpkg = $1;
                register $package, $VERSION, $dist;
                undef $VERSION;
                $package = $nextpkg;
            }
            else
            {   $package = $1;
            }

            trace "pkg $package from ".$file->name;
        }

        if( m/^ (?:use\s+version\s*;\s*)?
            (?:our)? \s* \$ ((?: \w+\:\:)*) VERSION \s* \= (.*)/x )
        {   defined $2 or next;
            my ($ns, $vers) = ($1, $2);

            # some versions of CPAN.pm do contain lines like "$VERSION =~ ..."
            # which also need to be processed.
            eval "\$VERSION =$vers";
            if(defined $VERSION)
            {   ($package = $ns) =~ s/\:\:$//
                    if length $ns;
                trace "pkg $package version $VERSION";
            }
        }
    }

    $VERSION = $VERSION->numify if ref $VERSION;
    register $package, $VERSION, $dist
        if defined $package;
}

sub update_global_cpan($$)
{   my ($mycpan, $globalcpan) = @_;

    my $global = catdir $mycpan, 'global';
    my ($mailrc, $globdetails, $modlist) = 
       map { catfile $global, $_ }
         qw/01mailrc.txt.gz 02packages.details.txt.gz 03modlist.data.gz/;

    return $globdetails
       if -f $globdetails && -f $globdetails && -f $modlist
       && -M $globdetails < $cpan_update;

    info "(re)loading global CPAN files";

    mkdirhier $global;
    load_file "$globalcpan/authors/01mailrc.txt.gz",   $mailrc;
    load_file "$globalcpan/modules/02packages.details.txt.gz", $globdetails;
    load_file "$globalcpan/modules/03modlist.data.gz", $modlist;
    $globdetails;
}

sub load_file($$)
{   my ($from, $to) = @_;
    my $response = $ua->get($from, ':content_file' => $to);
    return if $response->is_success;

    unlink $to;
    error __x"failed to get {uri} for {to}: {err}"
      , uri => $from, to => $to, err => $response->status_line;
}

sub merge_global_cpan($$$)
{   my ($mycpan, $pkgs, $globdetails) = @_;

    trace "merge packages with CPAN core list in $globdetails";
    my $cpan_pkgs = read_details $globdetails;

    while(my ($cpandist, $cpanpkgs) = each %$cpan_pkgs)
    {   foreach (@$cpanpkgs)
        {  my ($pkg, $version) = @$_;
           next if exists $pkgs->{$pkg};
           $pkgs->{$pkg} = [$version, $cpandist];
        }
    }
}

sub create_details($$$$$)
{  my ($details, $filename, $pkgs, $lazy, $undefs) = @_;

   trace "creating package details in $filename";
   my $fh = IO::Zlib->new($filename, 'wb')
      or fault __x"generating gzipped {fn}", fn => $filename;

   my $lines = keys %$pkgs;
   my $date  = time2str time;
   my $how   = $lazy ? "lazy" : "full";

   info "produced list of $lines packages $how";

   my $program     = basename $0;
   my $module      = __PACKAGE__;
   $fh->print (<<__HEADER);
File:         02packages.details.txt
URL:          file://$details
Description:  Packages listed in CPAN and local repository
Columns:      package name, version, path
Intended-For: private CPAN
Line-Count:   $lines
Written-By:   $program with $module $CPAN::Site::Index::VERSION ($how)
Last-Updated: $date

__HEADER

   foreach my $pkg (sort keys %$pkgs)
   {  my ($version, $path) = @{$pkgs->{$pkg}};

      $version = 'undef'
          if !defined $version || $version eq '';

      next
          if $version eq 'undef' && !$undefs;

      $path    =~ s,\\,/,g;
      $fh->printf("%-30s\t%s\t%s\n", $pkg, $version, $path);
   }
}

sub calculate_checksums($)
{   my $dirs = shift;
    trace "updating checksums";

    foreach my $dir (keys %$dirs)
    {   trace "summing $dir";
        CPAN::Checksums::updatedir($dir)
            or warning 'failed calculating checksums in {dir}', dir => $dir;
    }
}

sub read_details($)
{   my $fn = shift;
    -f $fn or return {};
    trace "collecting all details from $fn";

    my $fh    = IO::Zlib->new($fn, 'rb')
       or fault __x"cannot read from {fn}", fn => $fn;

    my $line;   # skip header, search first blank
    do { $line = $fh->getline } until $line =~ m/^\s*$/;

    my $time_last_update = (stat $fn)[9];
    my %dists;

    while(my $line = $fh->getline)
    {   chomp $line;
        my ($pkg, $version, $dist) = split ' ', $line, 3;

        unless($dist)
        {   warning "$fn error line=\n  $line";
            next;
        }

        push @{$dists{$dist}}, [$pkg, $version];
    }

    \%dists;
}

sub remove_expired_details($$$)
{   my ($mycpan, $dists, $newer) = @_;
    trace "extracting only existing local distributions";

    my $authors = catdir $mycpan, 'authors', 'id';
    foreach my $dist (keys %$dists)
    {   my $fn = catfile $authors, $dist;
        if(! -f $fn)
        {   # removed local or a global dist
            delete $dists->{$dist};
        }
        elsif(-M $fn < $newer)
        {   trace "dist $dist file updated, reindexing";
            delete $dists->{$dist};
        }
    }
}

sub mkdirhier(@)
{   foreach my $dir (@_)
    {   next if -d $dir;
        mkdirhier dirname $dir;

        mkdir $dir, 0755
            or fault __x"cannot create directory {dir}", dir => $dir;

        trace "created $dir";
    }
    1;
}

sub cpan_mirror($$$@)
{   my ($mycpan, $globalcpan, $mods, %opts) = @_;
    @$mods or return;
    my %need = map { ($_ => 1) } @$mods;
    my $auth = catdir $mycpan, 'authors', 'id';

    unless($ua)
    {   $ua = LWP::UserAgent->new;
        $ua->env_proxy if $opts{env_proxy};
    }

    my $globdetails
             = update_global_cpan $mycpan, $globalcpan;

    my $fh   = IO::Zlib->new($globdetails, 'rb')
        or fault __x"cannot read from {fn}", fn => $globdetails;

    while(my $line = $fh->getline)   # skip header, search first blank
    {   last if $line =~ m/^\s*$/;
    }

    $ua ||= LWP::UserAgent->new;
    while(my $line = $fh->getline)
    {   my ($pkg, $version, $dist) = split ' ', $line;
        delete $need{$pkg} or next;

        my $to = catfile $auth, split m#/#, $dist;
        if(-f $to)
        {   trace __x"package {pkg} present in distribution {dist}"
              , pkg => $pkg, dist => $dist;
            next;
        }

        my $source   = "$globalcpan/authors/id/$dist";
        mkdirhier dirname $to;
        my $response = $ua->get($source, ':content_file' => $to);
        unless($response->is_success)
        {   unlink $to;
            error __x"failed to get {uri} for {to}: {err}"
              , uri => $source, to => $to, err => $response->status_line;
        }

        info __x"got {pkg} in {dist}", pkg => $pkg, dist => $dist;
    }

    warning __x"package {pkg} does not exist", pkg => $_
        for sort keys %need;
}

1;


Makefile.PL
use ExtUtils::MakeMaker;

# Avoid the loading of a huge number of modules on clients.  When
# someone runs the script as client, it will get a problem with
# missing "CPAN::Checksums", which is not a problem.

eval "require CPAN::Site";
if(!$@ && $CPAN::Site::VERSION <= 1.01 && !$ENV{MARKOV_DEVEL})
{   warn <<__WARN;

*** the interface of CPAN::Site has changed a lot with release 1.01
*** please re-read the documentation.

__WARN
    sleep 5
}

WriteMakefile
 ( NAME       => 'CPAN::Site'
 , ABSTRACT   => 'adding site local modules'
 , VERSION    => '1.06'

 , EXE_FILES  => [ 'bin/cpansite' ]

 , AUTHOR     => 'Mark Overmeer'
 , PREREQ_PM  =>
   { Archive::Tar       => 1.00
   , CPAN::Checksums    => 0
   , Getopt::Long       => 0
   , HTTP::Date         => 0
   , IO::Zlib           => 0
   , Log::Report        => 0.25
   , LWP                => 0
   , Test::More         => 0.82
   , Test::Pod          => 1.00
   , version            => 0.76
   }
 , LICENSE    => 'perl'
 );

#### the next lines are added for OODoc, which generates the
#### distribution.
sub MY::postamble { <<__POSTAMBLE }

# for DIST
RAWDIR          = ../public_html/cpan-site/raw
DISTDIR         = ../public_html/cpan-site/source
LICENSE         = artistic

# for POD
FIRST_YEAR      = 1998,2005
EMAIL           = perl\@overmeer.net
WEBSITE         = http://perl.overmeer.net/cpan-site/

# for HTML
#HTML_OUTPUT	= ../public_html/cpan-site/html
#HTML_DOCROOT	= /cpan-site/html
__POSTAMBLE
